// vscode-languageclient/lib/protocolConverter.js
/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
'use strict';
var code = require('vscode');
var proto = require('./protocol');
var is = require('./utils/is');
var protocolCompletionItem_1 = require('./protocolCompletionItem');
function asDiagnostics(diagnostics) {
    return diagnostics.map(function (diagnostic) { return new code.Diagnostic(asRange(diagnostic.range), diagnostic.message, asDiagnosticSeverity(diagnostic.severity)); });
}
exports.asDiagnostics = asDiagnostics;
function asRange(value) {
    if (is.undefined(value)) {
        return undefined;
    }
    else if (is.nil(value)) {
        return null;
    }
    return new code.Range(asPosition(value.start), asPosition(value.end));
}
exports.asRange = asRange;
function asPosition(value) {
    if (is.undefined(value)) {
        return undefined;
    }
    else if (is.nil(value)) {
        return null;
    }
    return new code.Position(value.line, value.character);
}
exports.asPosition = asPosition;
function asDiagnosticSeverity(value) {
    switch (value) {
        case proto.DiagnosticSeverity.Error:
            return code.DiagnosticSeverity.Error;
        case proto.DiagnosticSeverity.Warning:
            return code.DiagnosticSeverity.Warning;
        case proto.DiagnosticSeverity.Information:
            return code.DiagnosticSeverity.Information;
        case proto.DiagnosticSeverity.Hint:
            return code.DiagnosticSeverity.Hint;
    }
}
exports.asDiagnosticSeverity = asDiagnosticSeverity;
function asHover(hover) {
    return new code.Hover(hover.contents, is.defined(hover.range) ? asRange(hover.range) : undefined);
}
exports.asHover = asHover;
function asCompletionItems(items) {
    return items.map(asCompletionItem);
}
exports.asCompletionItems = asCompletionItems;
function set(value, func) {
    if (is.defined(value)) {
        func();
    }
}
function asCompletionItem(item) {
    var result = new protocolCompletionItem_1.default(item.label);
    set(item.detail, function () { return result.detail = item.detail; });
    set(item.documentation, function () { return result.documentation = item.documentation; });
    set(item.filterText, function () { return result.filterText = item.filterText; });
    set(item.insertText, function () { return result.insertText = item.insertText; });
    // Protocol item kind is 1 based, codes item kind is zero based.
    set(item.kind, function () { return result.kind = item.kind - 1; });
    set(item.sortText, function () { return result.sortText = item.sortText; });
    set(item.textEdit, function () { return result.textEdit = asTextEdit(item.textEdit); });
    set(item.data, function () { return result.data = item.data; });
    return result;
}
exports.asCompletionItem = asCompletionItem;
function asTextEdit(edit) {
    return new code.TextEdit(asRange(edit.range), edit.newText);
}
exports.asTextEdit = asTextEdit;
function asSignatureHelp(item) {
    var result = new code.SignatureHelp();
    set(item.activeParameter, function () { return result.activeParameter = item.activeParameter; });
    set(item.activeSignature, function () { return result.activeSignature = item.activeSignature; });
    set(item.signatures, function () { return result.signatures = asSignatureInformations(item.signatures); });
    return result;
}
exports.asSignatureHelp = asSignatureHelp;
function asSignatureInformations(items) {
    return items.map(asSignatureInformation);
}
exports.asSignatureInformations = asSignatureInformations;
function asSignatureInformation(item) {
    var result = new code.SignatureInformation(item.label);
    set(item.documentation, function () { return result.documentation = item.documentation; });
    set(item.parameters, function () { return result.parameters = asParameterInformations(item.parameters); });
    return result;
}
exports.asSignatureInformation = asSignatureInformation;
function asParameterInformations(item) {
    return item.map(asParameterInformation);
}
exports.asParameterInformations = asParameterInformations;
function asParameterInformation(item) {
    var result = new code.ParameterInformation(item.label);
    set(item.documentation, function () { return result.documentation = item.documentation; });
    return result;
}
exports.asParameterInformation = asParameterInformation;
function asDefinitionResult(item) {
    if (is.array(item)) {
        return item.map(asLocation);
    }
    else {
        return asLocation(item);
    }
}
exports.asDefinitionResult = asDefinitionResult;
function asLocation(item) {
    return new code.Location(code.Uri.parse(item.uri), asRange(item.range));
}
exports.asLocation = asLocation;
function asReferences(values) {
    return values.map(asLocation);
}
exports.asReferences = asReferences;
function asDocumentHighlights(values) {
    return values.map(asDocumentHighlight);
}
exports.asDocumentHighlights = asDocumentHighlights;
function asDocumentHighlight(item) {
    var result = new code.DocumentHighlight(asRange(item.range));
    set(item.kind, function () { return result.kind = asDocumentHighlightKind(item.kind); });
    return result;
}
exports.asDocumentHighlight = asDocumentHighlight;
function asDocumentHighlightKind(item) {
    switch (item) {
        case proto.DocumentHighlightKind.Text:
            return code.DocumentHighlightKind.Text;
        case proto.DocumentHighlightKind.Read:
            return code.DocumentHighlightKind.Read;
        case proto.DocumentHighlightKind.Write:
            return code.DocumentHighlightKind.Write;
    }
    return code.DocumentHighlightKind.Text;
}
exports.asDocumentHighlightKind = asDocumentHighlightKind;
function asSymbolInformations(values, uri) {
    return values.map(function (information) { return asSymbolInformation(information, uri); });
}
exports.asSymbolInformations = asSymbolInformations;
function asSymbolInformation(item, uri) {
    // Symbol kind is one based in the protocol and zero based in code.
    var result = new code.SymbolInformation(item.name, item.kind - 1, asRange(item.location.range), item.location.uri ? code.Uri.parse(item.location.uri) : uri);
    set(item.containerName, function () { return result.containerName = item.containerName; });
    return result;
}
exports.asSymbolInformation = asSymbolInformation;
