// stylelint/dist/rules/media-feature-colon-space-before/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _utils = require("../../utils");

var _mediaFeatureColonSpaceAfter = require("../media-feature-colon-space-after");

var ruleName = "media-feature-colon-space-before";

exports.ruleName = ruleName;
var messages = (0, _utils.ruleMessages)(ruleName, {
  expectedBefore: function expectedBefore() {
    return "Expected single space before \":\"";
  },
  rejectedBefore: function rejectedBefore() {
    return "Unexpected whitespace before \":\"";
  }
});

exports.messages = messages;

exports["default"] = function (expectation) {
  var checker = (0, _utils.whitespaceChecker)("space", expectation, messages);
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, {
      actual: expectation,
      possible: ["always", "never"]
    });
    if (!validOptions) {
      return;
    }

    (0, _mediaFeatureColonSpaceAfter.mediaFeatureColonSpaceChecker)({
      root: root,
      result: result,
      locationChecker: checker.before,
      checkedRuleName: ruleName
    });
  };
};
