// stylelint/dist/rules/property-whitelist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _lodash = require("lodash");

var _postcss = require("postcss");

var _utils = require("../../utils");

var ruleName = "property-whitelist";

exports.ruleName = ruleName;
var messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: function rejected(p) {
    return "Unexpected property \"" + p + "\"";
  }
});

exports.messages = messages;

exports["default"] = function (whitelistInput) {
  var whitelist = [].concat(whitelistInput);
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, {
      actual: whitelist,
      possible: [_lodash.isString]
    });
    if (!validOptions) {
      return;
    }

    root.walkDecls(function (decl) {

      var prop = decl.prop;

      if ((0, _utils.cssPropertyIsVariable)(prop)) {
        return;
      }

      if (whitelist.indexOf(_postcss.vendor.unprefixed(prop)) === -1) {
        (0, _utils.report)({
          message: messages.rejected(prop),
          node: decl,
          result: result,
          ruleName: ruleName
        });
      }
    });
  };
};
