// stylelint/dist/rules/block-no-empty/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _utils = require("../../utils");

var ruleName = "block-no-empty";

exports.ruleName = ruleName;
var messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: "Unexpected empty block"
});

exports.messages = messages;

exports["default"] = function (actual) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, { actual: actual });
    if (!validOptions) {
      return;
    }

    // Check both kinds of statements: rules and at-rules
    root.walkRules(check);
    root.walkAtRules(check);

    function check(statement) {
      if ((0, _utils.cssStatementHasEmptyBlock)(statement)) {
        (0, _utils.report)({
          message: messages.rejected,
          node: statement,
          index: (0, _utils.cssStatementStringBeforeBlock)(statement, { noBefore: true }).length,
          result: result,
          ruleName: ruleName
        });
      }
    }
  };
};
