// stylelint/dist/rules/color-hex-case/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _utils = require("../../utils");

var ruleName = "color-hex-case";

exports.ruleName = ruleName;
var messages = (0, _utils.ruleMessages)(ruleName, {
  expected: function expected(h, v) {
    return "Expected \"" + h + "\" to be \"" + v + "\"";
  }
});

exports.messages = messages;

exports["default"] = function (expectation) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, {
      actual: expectation,
      possible: ["lower", "upper"]
    });
    if (!validOptions) {
      return;
    }

    root.walkDecls(function (decl) {
      var declString = decl.toString();
      (0, _utils.styleSearch)({ source: declString, target: "#" }, function (match) {

        var hexMatch = /^#[0-9A-Za-z]+/.exec(declString.substr(match.startIndex));
        if (!hexMatch) {
          return;
        }
        var hexValue = hexMatch[0];
        var hexValueLower = hexValue.toLowerCase();
        var hexValueUpper = hexValue.toUpperCase();

        if (expectation === "lower" && hexValue === hexValueLower) {
          return;
        }

        if (expectation === "upper" && hexValue === hexValueUpper) {
          return;
        }

        var variant = expectation === "lower" ? hexValueLower : hexValueUpper;

        (0, _utils.report)({
          message: messages.expected(hexValue, variant),
          node: decl,
          index: match.startIndex,
          result: result,
          ruleName: ruleName
        });
      });
    });
  };
};
