// stylelint/dist/rules/unit-blacklist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _lodash = require("lodash");

var _postcssValueParser = require("postcss-value-parser");

var _postcssValueParser2 = _interopRequireDefault(_postcssValueParser);

var _utils = require("../../utils");

var ruleName = "unit-blacklist";

exports.ruleName = ruleName;
var messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: function rejected(u) {
    return "Unexpected unit \"" + u + "\"";
  }
});

exports.messages = messages;

exports["default"] = function (blacklistInput) {
  var blacklist = [].concat(blacklistInput);
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, {
      actual: blacklist,
      possible: [_lodash.isString]
    });
    if (!validOptions) {
      return;
    }

    root.walkDecls(function (decl) {
      var value = decl.value;

      (0, _postcssValueParser2["default"])(value).walk(function (node) {
        var unit = _postcssValueParser2["default"].unit(node.value).unit;

        if (blacklist.indexOf(unit) !== -1 && node.type !== "string") {
          (0, _utils.report)({
            message: messages.rejected(unit),
            node: decl,
            index: (0, _utils.declarationValueIndexOffset)(decl) + node.sourceIndex,
            result: result,
            ruleName: ruleName
          });
        }
      });
    });
  };
};
