// stylelint/dist/rules/color-no-named/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _postcssValueParser = require("postcss-value-parser");

var _postcssValueParser2 = _interopRequireDefault(_postcssValueParser);

var _isCssColorName = require("is-css-color-name");

var _isCssColorName2 = _interopRequireDefault(_isCssColorName);

var _utils = require("../../utils");

var ruleName = "color-no-named";

exports.ruleName = ruleName;
var messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: function rejected(c) {
    return "Unexpected named color \"" + c + "\"";
  }
});

exports.messages = messages;

exports["default"] = function (actual) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, { actual: actual });
    if (!validOptions) {
      return;
    }

    root.walkDecls(function (decl) {
      var value = decl.value;

      (0, _postcssValueParser2["default"])(value).walk(function (node) {

        if ((0, _isCssColorName2["default"])(node.value) && !node.quote) {
          (0, _utils.report)({
            message: messages.rejected(node.value),
            node: decl,
            index: (0, _utils.declarationValueIndexOffset)(decl) + node.sourceIndex,
            result: result,
            ruleName: ruleName
          });
        }
      });
    });
  };
};
