// stylelint/dist/rules/no-eol-whitespace/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _utils = require("../../utils");

var ruleName = "no-eol-whitespace";

exports.ruleName = ruleName;
var messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: "Unexpected whitespace at end of line"
});

exports.messages = messages;
var whitespacesToReject = [" ", "\t"];

exports["default"] = function (actual) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, { actual: actual });
    if (!validOptions) {
      return;
    }

    var rootString = root.source.input.css;
    (0, _utils.styleSearch)({ source: rootString, target: ["\n", "\r"], checkComments: true }, function (match) {
      if (whitespacesToReject.indexOf(rootString[match.startIndex - 1]) !== -1) {
        (0, _utils.report)({
          message: messages.rejected,
          node: root,
          index: match.startIndex - 1,
          result: result,
          ruleName: ruleName
        });
      }
    });
  };
};
