// stylelint/dist/rules/comment-empty-line-before/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _utils = require("../../utils");

var ruleName = "comment-empty-line-before";

exports.ruleName = ruleName;
var messages = (0, _utils.ruleMessages)(ruleName, {
  expected: "Expected empty line before comment",
  rejected: "Unexpected empty line before comment"
});

exports.messages = messages;

exports["default"] = function (expectation, options) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, {
      actual: expectation,
      possible: ["always", "never"]
    }, {
      actual: options,
      possible: {
        except: ["first-nested"]
      },
      optional: true
    });
    if (!validOptions) {
      return;
    }

    root.walkComments(function (comment) {

      // Ignore the first node
      if (comment === root.first) {
        return;
      }

      if (comment.raws.inline) {
        return;
      }

      var before = comment.raw("before");

      // Ignore inline comments
      if (before.indexOf("\n") === -1) {
        return;
      }

      var expectEmptyLineBefore = (function () {
        if ((0, _utils.optionsHaveException)(options, "first-nested") && comment.parent !== root && comment === comment.parent.first) {
          return false;
        }
        return expectation === "always";
      })();

      var hasEmptyLineBefore = before.indexOf("\n\n") !== -1 || before.indexOf("\r\n\r\n") !== -1;

      // Return if the exceptation is met
      if (expectEmptyLineBefore === hasEmptyLineBefore) {
        return;
      }

      var message = expectEmptyLineBefore ? messages.expected : messages.rejected;

      (0, _utils.report)({
        message: message,
        node: comment,
        result: result,
        ruleName: ruleName
      });
    });
  };
};
