// stylelint/dist/rules/string-quotes/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _utils = require("../../utils");

var ruleName = "string-quotes";

exports.ruleName = ruleName;
var messages = (0, _utils.ruleMessages)(ruleName, {
  expected: function expected(q) {
    return "Expected " + q + " quotes";
  }
});

exports.messages = messages;

exports["default"] = function (expectation) {

  var erroneousQuote = expectation === "single" ? "\"" : "'";

  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, {
      actual: expectation,
      possible: ["single", "double"]
    });
    if (!validOptions) {
      return;
    }

    var cssString = root.source.input.css;
    (0, _utils.styleSearch)({ source: cssString, target: erroneousQuote }, function (match) {
      (0, _utils.report)({
        message: messages.expected(expectation),
        node: root,
        index: match.startIndex,
        result: result,
        ruleName: ruleName
      });
    });
  };
};
